File: packages/deep-engine/src/rayTracing/worldChunkBridge.ts
/**
 * 世界分区 → chunk 流送桥（波次4 前奏）：把世界单元流送计划翻译成 residency 域的
 * chunk 键集合合同。纯函数、确定性；真实 residency 挂载（AuthorChunkStream/
 * SceneChunkResidency.update 的 desired set）由接线切片消费本模块输出。
 * 键约定：世界单元 chunk 的 key 形如 `world|<cx>|<cz>|<lod>`——`world|` 前缀保证与
 * 作者场景 chunk 键空间永不冲突（authorChunkStream 以场景几何 id 命名）。
 */

import { worldCellCenter, worldCellKey, worldStreamingPlan, WORLD_CELL_LOD_COUNT,
  type WorldCellId, type WorldStreamingInput } from "./worldPartition.js";

export interface WorldChunkDemand {
  readonly key: string;
  readonly cell: WorldCellId;
  readonly lod: number;
  /** 0..1，越小越紧迫（与 worldStreamingPriority 同序归一）。 */
  readonly urgency: number;
  readonly center: readonly [number, number, number];
}

export interface WorldChunkBridgeOptions {
  /** 同时处于期望集的单元数上限（含各 LOD 层）。 */
  readonly maxCells?: number;
  /** 期望集中保留的 LOD 层数（0 = 仅最细层）。默认全部。 */
  readonly maxLod?: number;
}

const DEFAULT_MAX_CELLS = 64;

/** 相机流送输入 → 有序 chunk 期望集（紧迫度升序；同紧迫度 key 字典序）。 */
export function planWorldChunkDemand(input: WorldStreamingInput, radius: number,
  options: WorldChunkBridgeOptions = {}): readonly WorldChunkDemand[] {
  const maxCells = options.maxCells ?? DEFAULT_MAX_CELLS;
  if (!Number.isSafeInteger(maxCells) || maxCells < 1) throw new RangeError("maxCells must be a positive integer.");
  const maxLod = options.maxLod ?? WORLD_CELL_LOD_COUNT;
  if (!Number.isSafeInteger(maxLod) || maxLod < 1 || maxLod > WORLD_CELL_LOD_COUNT) {
    throw new RangeError(`maxLod must be an integer in [1,${WORLD_CELL_LOD_COUNT}].`);
  }
  const plan = worldStreamingPlan(input, radius).slice(0, maxCells);
  // 紧迫度按流送优先级秩次归一到 0..1（秩 0 = 最优先，即相机单元），与
  // worldStreamingPriority 严格同序；同一单元各 LOD 共享同一紧迫度。
  // 秩次归一对任意半径/maxCells 组合都确定且有界，供消费方做 visible/prefetch 分档。
  const rankDenominator = Math.max(1, plan.length - 1);
  const cells: WorldChunkDemand[] = [];
  for (let index = 0; index < plan.length; index++) {
    const cell = plan[index]!;
    const urgency = index / rankDenominator;
    const center2d = worldCellCenter(cell);
    for (let lod = 0; lod < maxLod; lod++) {
      cells.push({ key: worldChunkKey(cell, lod), cell, lod, urgency,
        center: [center2d.x, 0, center2d.z] });
    }
  }
  return cells;
}

export function worldChunkKey(cell: WorldCellId, lod: number): string {
  return `world|${worldCellKey(cell)}|lod${lod}`;
}

/** 键解析：非 world| 前缀返回 undefined（作者 chunk 键不受影响）。 */
export function parseWorldChunkKey(key: string): { cell: WorldCellId; lod: number } | undefined {
  const parts = key.split("|");
  if (parts.length !== 4 || parts[0] !== "world") return undefined;
  const cx = Number(parts[1]!), cz = Number(parts[2]!);
  const lodMatch = /^lod(\d+)$/.exec(parts[3]!);
  if (!Number.isSafeInteger(cx) || !Number.isSafeInteger(cz) || !lodMatch) return undefined;
  const lod = Number(lodMatch[1]);
  if (!Number.isSafeInteger(lod) || lod < 0 || lod >= WORLD_CELL_LOD_COUNT) return undefined;
  return { cell: { cx, cz }, lod };
}

File: packages/deep-engine-native/src/world_chunk_bridge.rs
//! 世界分区 → chunk 流送需求桥的 Native 镜像（波次4）。
//! 与 TS `rayTracing/worldChunkBridge.ts` 逐语义对齐：`world|cx|cz|lodN` 键空间与
//! 作者场景 chunk 键永不冲突；maxCells/maxLod 有界；解析 fail-closed。

use crate::world_partition::{
    WORLD_CELL_LOD_COUNT, WorldCellId, WorldStreamingInput, world_streaming_plan,
};

pub const DEFAULT_MAX_CELLS: u32 = 64;

#[derive(Debug, Clone, PartialEq)]
pub struct WorldChunkDemand {
    pub key: String,
    pub cell: WorldCellId,
    pub lod: u32,
    pub center: (f64, f64),
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct WorldChunkBridgeOptions {
    pub max_cells: u32,
    /// 期望集中保留的 LOD 层数（1..=WORLD_CELL_LOD_COUNT）。
    pub max_lod: u32,
}

pub fn world_chunk_key(cell: WorldCellId, lod: u32) -> String {
    format!("world|{}|{}|lod{}", cell.cx, cell.cz, lod)
}

/// 键解析：非 `world|` 前缀（作者 chunk 键）或 lod 越界返回 None。
pub fn parse_world_chunk_key(key: &str) -> Option<(WorldCellId, u32)> {
    let parts: Vec<&str> = key.split('|').collect();
    if parts.len() != 4 || parts[0] != "world" {
        return None;
    }
    let cx = parts[1].parse::<i64>().ok()?;
    let cz = parts[2].parse::<i64>().ok()?;
    let lod = parts[3].strip_prefix("lod")?.parse::<u32>().ok()?;
    if lod >= WORLD_CELL_LOD_COUNT {
        return None;
    }
    Some((WorldCellId { cx, cz }, lod))
}

/// 相机流送输入 → lod 键控 chunk 期望集（与 TS planWorldChunkDemand 同序同界）。
pub fn plan_world_chunk_demand(
    input: &WorldStreamingInput,
    radius: i64,
    options: WorldChunkBridgeOptions,
) -> Result<Vec<WorldChunkDemand>, String> {
    if options.max_cells == 0 {
        return Err("maxCells must be a positive integer.".to_string());
    }
    if options.max_lod == 0 || options.max_lod > WORLD_CELL_LOD_COUNT {
        return Err(format!(
            "maxLod must be an integer in [1,{WORLD_CELL_LOD_COUNT}]."
        ));
    }
    let plan = world_streaming_plan(input, radius)?;
    let mut demands =
        Vec::with_capacity(plan.len().min(options.max_cells as usize) * options.max_lod as usize);
    for entry in plan.into_iter().take(options.max_cells as usize) {
        let center = crate::world_partition::world_cell_center(entry.cell);
        for lod in 0..options.max_lod {
            demands.push(WorldChunkDemand {
                key: world_chunk_key(entry.cell, lod),
                cell: entry.cell,
                lod,
                center,
            });
        }
    }
    Ok(demands)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn expands_cells_into_lod_keyed_demands_with_bounds() {
        let input = WorldStreamingInput {
            camera_x: 900.0,
            camera_z: -700.0,
            forward_x: 1.0,
            forward_z: 0.0,
            speed: 10.0,
        };
        let demands = plan_world_chunk_demand(
            &input,
            1,
            WorldChunkBridgeOptions {
                max_cells: 9,
                max_lod: 2,
            },
        )
        .unwrap();
        assert_eq!(demands.len(), 18);
        assert!(demands.iter().all(|demand| demand.lod < 2));
        let unique: std::collections::HashSet<_> =
            demands.iter().map(|demand| demand.key.clone()).collect();
        assert_eq!(unique.len(), 18);
    }

    #[test]
    fn round_trips_keys_and_rejects_author_keys() {
        let key = world_chunk_key(WorldCellId { cx: -3, cz: 12 }, 2);
        assert_eq!(
            parse_world_chunk_key(&key),
            Some((WorldCellId { cx: -3, cz: 12 }, 2))
        );
        assert_eq!(parse_world_chunk_key("some-scene-geometry|deep-vg-0"), None);
        assert_eq!(parse_world_chunk_key("world|1|2|lod9"), None);
        assert_eq!(parse_world_chunk_key("world|1|2|core"), None);
    }

    #[test]
    fn validates_options_fail_closed() {
        let input = WorldStreamingInput {
            camera_x: 0.0,
            camera_z: 0.0,
            forward_x: 0.0,
            forward_z: 0.0,
            speed: 0.0,
        };
        assert!(
            plan_world_chunk_demand(
                &input,
                1,
                WorldChunkBridgeOptions {
                    max_cells: 0,
                    max_lod: 1
                }
            )
            .is_err()
        );
        assert!(
            plan_world_chunk_demand(
                &input,
                1,
                WorldChunkBridgeOptions {
                    max_cells: 4,
                    max_lod: 5
                }
            )
            .is_err()
        );
    }
}
